PERF: Improve pickle support with BZ2 & LZMA

jakirkham opened the pull request:
- [x] Fixes #46747
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

jakirkham committed:
Add `LZMAFile` wrapper for pickle protocol 5

jakirkham committed:
Add `BZ2File` wrapper for pickle protocol 5

jakirkham committed:
Use BZ2 & LZMA wrappers for full pickle support

jakirkham committed:
Workaround linter issue

`PickleBuffer` isn't currently included in `SupportBytes`, which causes
issues with pyright when passing `PickleBuffer` instances to `bytes`.
Though it appears ok passing `PickleBuffer` instances to `memoryview`s.
So do that instead. This is functionaly very equivalent. There is a
slight performance cost to making a `memoryview`, but this is likely
negligible compared to copying to `bytes`.

mroeschke commented:
Noting that we're having unrelated, CI issues currently. Hopefully it should be fixed today.

jakirkham commented:
No worries. Thanks Matthew! 🙏

Looks like any related test failures have been fixed, but will recheck once CI issues clear.

If you have thoughts on this approach, please let me know. Though no worries if you are busy 🙂

mroeschke commented:
The approach seems reasonable; it would be good to ensure we have tests to ensure the `raw` and `memoryview` path are both tested.

Also am I correct in understanding in https://github.com/pandas-dev/pandas/issues/46747 that once we drop 3.9 that this won't be entirely necessary or would it still be a good performance benefit to keep?

jakirkham commented:
Mostly, [Python 3.9.6 has the fix]( https://docs.python.org/3.9/whatsnew/changelog.html?highlight=bpo-44439#python-3-9-6-final ). Also the fix is included in [Python 3.10.0b4]( https://docs.python.org/3.10/whatsnew/changelog.html?highlight=bpo-44439#python-3-10-0-beta-4 ) and [Python 3.11.0a1]( https://docs.python.org/3.11/whatsnew/changelog.html?highlight=bpo-44439#python-3-11-0-alpha-1 ). So Python 3.9.6+ don't need the workaround. Python 3.8 and earlier versions of Python 3.9 would need the workaround. IOW it is probably safer to keep until Python 3.10 is the minimum.

Once Python 3.10 is minimum, we could drop the wrapper classes and just keep the simplified `to_pickle`.

Happy to flag the code based on Python version somehow if that makes it easier to track.

In terms of testing, think most of what we need is covered in [these tests]( https://github.com/pandas-dev/pandas/blob/864729813a0203af8bb0d30b6c883588ae2c96f8/pandas/tests/io/test_pickle.py#L256 ). The default path `raw` should be handled there (though please correct me if I'm missing something). The `memoryview` case would only come up if the data was not contiguous somehow (for example `numpy.arange(10)[::2]`). Do you know of good examples in Pandas where that would come up?

twoertwein commented on pandas/io/common.py:
```diff
@@ -1002,6 +1003,22 @@ def write_to_buffer(self) -> None:
         self.buffer.addfile(tarinfo, self)
 
 
+class _BZ2File(bz2.BZ2File):
```
It would be nice to share the method between `_BZ2File` and `_LZMAFile` - could have a mixin class but that would probably not work with the super call.

jakirkham replied:
Yeah was thinking the same thing. May actually break it out as a utility function, which could also aid in testing

jakirkham committed:
Refactor out `flatten_buffer`

jakirkham committed:
Refactor `B2File` into separate module

jakirkham committed:
Merge pandas-dev/main into jakirkham/fix_pickle5

jakirkham committed:
Test `flatten_buffer`

This provides a reasonable proxy for testing patched `BZ2File` and
`LZMAFile` objects.

jakirkham committed:
Move `flatten_buffer` to `_utils`

This ran into cyclic import issues in `pickle_compat`. So move
`flatten_buffer` to its own module free of these issues.

jakirkham committed:
Import `annotations` to fix `|` usage

jakirkham committed:
Merge pandas-dev/main into jakirkham/fix_pickle5

jakirkham committed:
Sort `import`s to fix lint

jakirkham committed:
Patch `BZ2File` & `LZMAFile` on Python pre-3.10

This should limit the effects of this patch. Also should make it easier
to remove this backport later once all supported Python versions have
the fix.

jakirkham committed:
Test C & F contiguous NumPy arrays

Also test another non-contiguous array.

jakirkham committed:
Test `memoryview` is 1-D `uint8` contiguous data

If a `memoryview` is returned, make sure it as close to `bytes` |
`bytearray` as possible. This ensures if other functions assume
something like `bytes` (for example assuming `len(b)` is the number of
bytes contained), things will continue to work even though this is a
`memoryview`.

jakirkham commented:
Sounds good. Have refactored out a common utility function, `flatten_buffer`, which simplifies both subclasses. Also `flatten_buffer` is tested in a variety of cases to make sure it is coercing data into something both compressors can handle (without the fix). Also conditioned the subclass patches on Python version support to clarify things. The subclasses themselves have been moved into `compat`.

twoertwein commented on pandas/compat/_utils.py:
```diff
@@ -0,0 +1,29 @@
+import pickle
+
+
+def flatten_buffer(b: bytes | bytearray | memoryview | pickle.PickleBuffer):
```
Great change!

I'm not sure whether there is a policy about structuring `pandas.compat` @mroeschke - might be nicer to have this function and the two classes in one private file (or at least rename the files `lzma.py` and `bz2.py` to be private).

jakirkham replied:
Thanks!

Yeah happy to restructure however is easiest.

Just was running into cyclic `ImportError`s so had to start breaking things out somehow (even if it would be changed later).

mroeschke replied:
Yeah agreed with @twoertwein; would be nice to keep `flatten_buffer`, `BZ2File`, and `LZMAFile` together in one file so it's easier to see they are all related. Could still be a separate file like `pandas/compat/_pickle_compat.py` or similar.

mroeschke replied:
Could the output of this function be typed as well?

jakirkham replied:
Consolidated into `_compressors.py`. Happy to iterate further as needed

jakirkham committed:
Run `black` on `bz2` and `lzma` compat files

jakirkham committed:
One more lint fix

jakirkham committed:
Drop unused `PickleBuffer` `import`s

mroeschke commented on pandas/compat/bz2.py:
```diff
@@ -0,0 +1,27 @@
+
+from pandas.compat._utils import flatten_buffer
+
+if sys.version_info < (3, 10):
```
Could you use `not PY310:` (PY310 in `pandas/compat/__init__.py`)

jakirkham replied:
Am afraid this will lead to more cyclic `ImportError`s as noted above. Not seeing an easy way to resolve this

jakirkham replied:
Tried just copying over the `PY310` definition verbatim. Hopefully that makes `grep`ping easier even though it is now defined in two places. Maybe we could refactor out into a constants file that is more free of cyclic import issues.

jakirkham committed:
Simplify change to `panda.compat.__init__`

Now that the LZMA changes are in a separate file, cleanup the changes to
`pandas.compat.__init__`.

jakirkham committed:
Type `flatten_buffer` result

jakirkham committed:
Use `order="A"` in `memoryview.tobytes(...)`

In the function `flatten_buffer`, the order is already effectively
enforced when copying can be avoided by using `PickleBuffer.raw(...)`.
However some test comparisons failed (when they shouldn't have) as this
wasn't specified. So add the `order` in both the function and the test.
This should fix that test failure.

jakirkham committed:
Move all compat compressors into a single file

jakirkham committed:
Fix `BZ2File` `import`

jakirkham committed:
Merge pandas-dev/main into jakirkham/fix_pickle5

jakirkham committed:
Refactor out common compat constants

jakirkham committed:
Fix `import` sorting

jakirkham committed:
Drop unused `import`

jakirkham committed:
Ignore `flake8` errors on wildcard `import`

jakirkham committed:
Revert "Ignore `flake8` errors on wildcard `import`"

This reverts commit f1f1a2e56861abbc0b2c04adbb98880794c47ca0.

jakirkham committed:
Explicitly `import` all constants

jakirkham committed:
Assign `IS64` first

jakirkham committed:
Try `noqa` on wildcard `import` again

jakirkham commented on pandas/compat/__init__.py:
```diff
@@ -12,9 +12,10 @@
 
 from pandas._typing import F
+import pandas.compat._compressors
+from pandas.compat._constants import *
```
Ok got `noqa` to work in the end.

jakirkham committed:
Declare `BZ2File` & `LZMAFile` once

Fixes a linter issue from pyright.

jakirkham committed:
`import PickleBuffer` for simplicity

jakirkham committed:
Add `bytearray` to return type

jakirkham committed:
Test `bytes` & `bytearray` are returned unaltered

jakirkham committed:
Merge branch 'main' into fix_pickle5

jakirkham commented:
Consolidate the compat changes into a single file. Also refactored `PY310` and other platform constants into a single file used here. Added more types. Also added a few more test cases. Please let me know what you think 🙂

jakirkham commented:
Tried to use the request a review feature, but it seems to not allow requesting both reviewers 🤷‍♂️

twoertwein commented on pandas/compat/__init__.py:
```diff
@@ -12,9 +12,10 @@
 
 from pandas._typing import F
+import pandas.compat._compressors
+from pandas.compat._constants import *  # noqa: F401, F403
```
Would prefer explicitly listening all the constants

jakirkham replied:
Sounds good. Done in the recent commit. Please take a look :)

twoertwein reviewed (approved):
One small change, lgtm otherwise

jakirkham committed:
Merge pandas-dev/main into jakirkham/fix_pickle5

jakirkham committed:
Explicit list all constants

jakirkham commented:
Explicitly listed all constants.

Have not added a news entry yet. Please let me know where we should add this.

Idk if 1.x is still being developed, but it may make sense to backport there as well if it is.

twoertwein commented on pandas/compat/__init__.py:
```diff
@@ -12,9 +12,16 @@
+    PY39,
+    PY310,
+    PY311,
+    PYPY,
```
Can you use the redundant export `IS64 as IS64`, then type checkers (and linters) know that these imports are meant to be re-exported

jakirkham replied:
Yep updated to do that instead. Thanks!

jakirkham replied:
Unfortunately `flake8` saw through this, but did an identity assignment, which seemed to fix it while still being explicit about imports. Please let me know what you think.

twoertwein replied:
Sorry, I mixed the policy of pandas-stubs and pandas up: pandas-stubs uses the redundant imports in pyi files; pandas re-exports imports by listing all impotable symbols in `__all__` (which is flake8-compatible)

At the bottom of the file, I would add `__all__ = ["PY39", ...]` unfortunately, that means you will also need to list other functions/classes defined in this file (not just those constants).

jakirkham replied:
Gotcha. Pushed that change in a recent commit.

jakirkham committed:
Trick linter into thinking constants are used ;)

jakirkham committed:
Merge pandas-dev/main into jakirkham/fix_pickle5

jakirkham committed:
Add new entry to 2.0.0

jakirkham commented on doc/source/whatsnew/v2.0.0.rst:
```diff
@@ -170,6 +170,7 @@ Performance improvements
 - Performance improvement in :func:`merge` and :meth:`DataFrame.join` when joining on a sorted :class:`MultiIndex` (:issue:`48504`)
 - Performance improvement in :meth:`DataFrame.loc` and :meth:`Series.loc` for tuple-based indexing of a :class:`MultiIndex` (:issue:`48384`)
 - Performance improvement for :meth:`MultiIndex.unique` (:issue:`48335`)
+- Performance use out-of-band pickling with BZ2 & LZMA files (:issue:`49068`)
```
How does this look?

jakirkham replied:
How about this?

```suggestion
- Minimize memory usage when writing BZ2 & LZMA files by using out-of-band pickling (:issue:`49068`)
```

jakirkham replied:
Went ahead and included this. Though tweaks are still welcome :)

jakirkham committed:
Assign constants to themselves

Should work around linter issues.

jakirkham commented on pandas/compat/__init__.py:
```diff
@@ -12,9 +12,16 @@
+    PY39 as PY39,
+    PY310 as PY310,
+    PY311 as PY311,
+    PYPY as PYPY,
```
Used an assignment trick to fix this issue

mroeschke commented on pandas/compat/__init__.py:
```diff
@@ -26,14 +33,11 @@
-PY311 = sys.version_info >= (3, 11)
-PYPY = platform.python_implementation() == "PyPy"
-IS64 = sys.maxsize > 2**32
+IS64 = IS64
```
I think if you define these in `__all__` as described in https://github.com/pandas-dev/pandas/pull/49068#discussion_r1000791860 this redundant assignment can be avoided

jakirkham replied:
Oh the assignment here is to appease the linter while explicitly `import`ing these. Please see these threads for context:

* https://github.com/pandas-dev/pandas/pull/49068#discussion_r1000051268
* https://github.com/pandas-dev/pandas/pull/49068#discussion_r1000073840

That said, happy to add to `__all__` if we want that

jakirkham committed:
Update changelog entry    [skip ci]

jakirkham committed:
Merge pandas-dev/main into jakirkham/fix_pickle5

jakirkham committed:
Add constants to `__all__`

jakirkham commented:
Anything else needed here?

twoertwein commented on doc/source/whatsnew/v2.0.0.rst:
```diff
@@ -185,6 +185,7 @@ Performance improvements
 - Performance improvement in :func:`merge` and :meth:`DataFrame.join` when joining on a sorted :class:`MultiIndex` (:issue:`48504`)
 - Performance improvement in :meth:`DataFrame.loc` and :meth:`Series.loc` for tuple-based indexing of a :class:`MultiIndex` (:issue:`48384`)
 - Performance improvement for :meth:`MultiIndex.unique` (:issue:`48335`)
+- Minimize memory usage when writing BZ2 & LZMA files by using out-of-band pickling (:issue:`49068`)
```
Would be good to reference DataFrame/Series.to_pickle in the whatsnew

jakirkham replied:
```suggestion
- Reduce memory usage of DataFrame/Series `to_pickle` when using BZ2 or LZMA (:issue:`49068`)
```

jakirkham committed:
Update changelog entry    [ci skip]

jakirkham commented:
How does the updated changelog entry look?

twoertwein commented on doc/source/whatsnew/v2.0.0.rst:
```diff
@@ -185,6 +185,7 @@ Performance improvements
 - Performance improvement in :func:`merge` and :meth:`DataFrame.join` when joining on a sorted :class:`MultiIndex` (:issue:`48504`)
 - Performance improvement in :meth:`DataFrame.loc` and :meth:`Series.loc` for tuple-based indexing of a :class:`MultiIndex` (:issue:`48384`)
 - Performance improvement for :meth:`MultiIndex.unique` (:issue:`48335`)
+- Reduce memory usage of DataFrame/Series `to_pickle` when using BZ2 or LZMA (:issue:`49068`)
```
Sorry for all the nitpicks, can you please use the patterns used elsewhere:
:meth:`DataFrame.to_pickle` and  :meth:`Series.to_pickle`

jakirkham replied:
```suggestion
- Reduce memory usage of :meth:`DataFrame.to_pickle`/:meth:`Series.to_pickle` when using BZ2 or LZMA (:issue:`49068`)
```

jakirkham committed:
Use Sphinx method annotation

jakirkham committed:
Merge pandas-dev/main into jakirkham/fix_pickle5

mroeschke commented:
Sweet thanks @jakirkham!

jakirkham commented:
Awesome! Thank you both 😄

Would it be possible to backport this change as well? If so, what is needed from me to do that?

mroeschke commented:
Backports are currently done for regressions (bug/performance) that were introduced in 1.5. IIUC the performance of `to_pickle` in this situation didn't necessarily regress in 1.5?

jakirkham commented:
Good question. The history goes like this:

* PR ( https://github.com/pandas-dev/pandas/pull/37056 ) enabled this for all compressors in 1.2
* Users discovered issues with BZ2 & LZMA ( https://github.com/pandas-dev/pandas/issues/39002 )
* PR ( https://github.com/pandas-dev/pandas/pull/39376 ) special cased BZ2 & LZMA to fix the bug (though this has reduced perf, which of course is the right tradeoff)

So the perf improvement and then reduction occurred in 1.2 (not 1.5).

That said, maybe there is still value in having this perf improvement in some 1.x version (especially given the move to 2 occurring). What do you think? 🙂

mroeschke commented:
Would be a nice performance benefit for 1.5.x users. Also need to weigh any unknown bugs/behavior changes this might introduce. Thoughts @pandas-dev/pandas-core?

WillAyd commented:
I would err on the side of caution and go 2.0

phofl commented:
I'd also prefer to keep this in on main only
